src/FirmwarePlugin/FirmwarePluginManager.cc

```diff
@@ -42,11 +42,46 @@ FirmwarePluginManager::~FirmwarePluginManager()
 
 }
 
-FirmwarePlugin* FirmwarePluginManager::firmwarePluginForAutopilot(MAV_AUTOPILOT autopilotType)
+FirmwarePlugin* FirmwarePluginManager::firmwarePluginForAutopilot(MAV_AUTOPILOT autopilotType, MAV_TYPE vehicleType)
 {
     switch (autopilotType) {
         case MAV_AUTOPILOT_ARDUPILOTMEGA:
-            return APMFirmwarePlugin::instance();
+            switch (vehicleType) {
+                case MAV_TYPE_QUADROTOR:
+                case MAV_TYPE_HEXAROTOR:
+                case MAV_TYPE_OCTOROTOR:
+                case MAV_TYPE_TRICOPTER:
+                    return ArduCopterFirmwarePlugin::instance();
+                    break;
+                // FIXME: The remainder of these need to be correctly assigned and new plugin classes created as needed.
```

Comment:
APMFirmwarePlugin is now the base class for all APM vehicle type (MAV_TYPE) specific firmware plugins. It is no longer a standalone firmware plugin. You need to implement ArduPlaneFirmwarePlugin, RoverFirmwarePlugin, ... plugins following the model of ArduCopterFirmwarePlugin. Then you are going to need to modify this switch to create the appropriate plugin according to MAV_TYPE. I've already hooked up ArduCopterFirmwarePlugin but I'm not sure I have all the MAV_TYPE's that should be associated with it. For example does MAV_TYPE_COAXIAL and MAV_TYPE_HELICOPTER go with it as well?